### PR TITLE
Rate limiter's configuration

### PR DESCRIPTION
**Pull request by Garnek20**
In this PR you'll find:
- configuration changes to specify 
    - `rate_limiter` per `user` and `cluster_user`
    - `redis_clients` to create redis clients 
- `redisFactory` to cache redis clients and avoid too many open connections
- updated tests to validate configuration
- updated configuration/README

### CODE REVIEW

File: README-CN.md [hunk 1]
```diff
@@ -207,11 +206,11 @@ clusters:
 
 caches:
   - name: "shortterm"
-    dir: "/path/to/cache/dir"
-    max_size: 150Mb
-
-    # Cached responses will expire in 130s.
-    expire: 130s
+    mode: "file_system"
```
Comment (Garnek20) [inline README-CN.md:209]: Since I was changing configuration, I updated obsolete Readmes

File: clients/redis.go [hunk 2]
```diff
@@ -7,7 +7,35 @@ import (
 	"github.com/go-redis/redis/v8"
 )
 
-func NewRedisClient(cfg config.RedisCacheConfig) (redis.UniversalClient, error) {
+type RedisFactory struct {
```
Comment (r-zenine) [inline clients/redis.go:10]: `RedisProxy` maybe ? 
Factory might not be the most telling name

File: config/config.go [hunk 3]
```diff
@@ -84,6 +87,32 @@ func withoutSensitiveInfo(config *Config) *Config {
...
+	Username  string                 `yaml:"username,omitempty"`
+	Password  string                 `yaml:"password,omitempty"`
+	Addresses []string               `yaml:"addresses"`
+
+	// Catches all undefined fields
+	XXX map[string]interface{} `yaml:",inline"`
+}
+
+func (r *RedisClient) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type plain RedisClient
```
Comment (r-zenine) [inline config/config.go:104]: you want a `var` not a `type` here correct ?

File: config/config_test.go [hunk 4]
```diff
@@ -112,10 +117,15 @@ func TestLoadConfig(t *testing.T) {
 								Name:                 "default",
 								MaxConcurrentQueries: 4,
 								MaxExecutionTime:     Duration(time.Minute),
+								RateLimiter: RateLimiterConfig{
+									RateLimiterAlgorithm: "redis.sliding-window",
```
Comment (r-zenine) [inline config/config_test.go:121]: Can we get that string in a const ?

File: scope.go [hunk 5]
```diff
@@ -214,17 +215,17 @@ func (s *scope) inc() error {
...
 		}
 
 	}
 
 	if s.clusterUser.rateLimiter != nil && err == nil {
-		if s.clusterUser.rateLimiter.Limit(limiter.RateLimitKey{Key: s.clusterUser.name, Limit: s.clusterUser.reqPerMin, Window: time.Minute}) {
-			err = fmt.Errorf("rate limit for cluster user %q is exceeded: requests_per_minute limit: %d",
-				s.clusterUser.name, s.clusterUser.reqPerMin)
+		if s.clusterUser.rateLimiter.Limit(limiter.RateLimitKey{Key: s.clusterUser.name, Limit: s.clusterUser.userRateLimiter.Limit, Window: s.clusterUser.userRateLimiter.WindowDuration}) {
+			err = fmt.Errorf("rate limit for cluster user %q is exceeded: requests per %s limit: %d",
```
Comment (r-zenine) [inline scope.go:227]: can you introduce a 
```
var ErrRateLimitReached = errors.New("........")
``` 
and use it here with `%w` in `fmt.Errorf`

File: clients/redis.go (hunk 2 above)
Comment (Garnek20) [reply clients/redis.go:10]: Makes sense https://refactoring.guru/design-patterns/proxy

File: config/config.go (hunk 3 above)
Comment (Garnek20) [reply config/config.go:104]: Actually not. It's the `chproxy`'s way to handle unmarshalling. 🤷  
If I change it I'll have to change all other `UnmarshalYAML` definitions

File: scope.go (hunk 5 above)
Comment (Garnek20) [reply scope.go:227]: Sure! I'll add it in the following PR #16 because it moves this part of code away.

File: config/config_test.go (hunk 4 above)
Comment (Garnek20) [reply config/config_test.go:121]: Well, it's actually a test and all other params are hardcoded so not sure it brings any value. WDYT?

**Pull request merged**